doc/BGP/BGP-router-id.md

```diff
@@ -0,0 +1,127 @@
+2. SONiC wouldn't add BGP peer when there is not Loopback0 IPv4 exists. It would cause BGP cannot establish.
+
+Below is current workflow about BGP and router id, only includes contents related to Loopback0.
+
+1. After bgp container started, configuration file `/etc/frr/bgpd.conf` for bgpd would be rendered. It will use Loopback0 IPv4 address as BGP router id, if it doesn't exist, the BGP router id wouldn't be specified.
+2. bgpd start with configuration rendered before. If BGP router id is not specified, it would choose an IP address in device to be BGP router id.
+3. After bgpcfgd started, it will add bgp peer depends on whether Loopback0 IPv4 exist.
+
+<p align=center>
+<img src="img/origin_bgp_seq.png" alt="Figure 1. Origin bgp seq" width=700>
+</p>
+
+### Requirements
+
+Add support to allow user explicitly configure BGP router id.
+
+### High Level Design
+
+2 aspects enhancement:
+
+1. Add a field `bgp_router_id` in `CONFIG_DB["DEVICE_METADATA"]["localhost"]` to support explicitly configure BGP router id. If `CONFIG_DB["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` configured, always use it as BGP router id no matter whether Loopback0 IPv4 exist. With this change, the new BGP router id configuration behavior will be like follow. To be clarified that when bgp_router_id doesn't be configured, the behavior is totally same as previously.
+
+|           | Loopback0 IPv4 address exists | Loopback0 IPv4 address doesn't exist |
+|--------------|-------|------------|
+| bgp_router_id configured | Honor bgp_router_id | Honor bgp_router_id |
+| bgp_router_id doesn't be configured | Honor Loopback0 IPv4 address | FRR default router ID value is selected as a IP Address of the device. When router zebra is not enabled bgpd can’t get interface information so router-id is set to 0.0.0.0 |
+
+2. Remove strong dependencies on Loopback0 IPv4 address when adding BGP peer in the situation that bgp_router_id is configured.With this change, the new BGP peer adding behavior will be like follow. To be clarified that when bgp_router_id doesn't be configured, the behavior is totally same as previously.
+
+|          | Loopback0 IPv4 address exists | Loopback0 IPv4 address doesn't exist |
+|--------------|-------|------------|
+| bgp_router_id configured | Add BGP peer | Add BGP peer |
+| bgp_router_id doesn't be configured | Add BGP peer | Do not add BGP peer |
+
+Below is new workflow, the main changes are in `1.` and `3.`.
+
+1. After bgp container started, configuration file `/etc/frr/bgpd.conf` for bgpd is would be rendered.
+   * If CONFIG_DB`["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` exists, use it as BGP router id.
+   * Else if Loopback0 IPv4 address exists, use it as BGP router id.
+   * Else, BGP router id wouldn't be specified.
```

Comment:
Should we break and report failure in the final "else" branch? Should we allow device working in this way?

Reply:
If we consider this case as critical failure, we may need to add constrain in yang model.

Reply:
This is existing behaviour. The design is trying to not make breaking change.

Existing behaviour:
bgpcfgd will not adding neighbors if router id not specified and bgp is actually not working.

Reply:
Existing behaviour does not always mean right behavior. If we consider this case as critical failure, we may need to add constrain in yang model.